lib/src/rules/diagnostic_describe_all_properties.dart

```diff
@@ -0,0 +1,248 @@
+* not themselves `Widget`s or collections of `Widget`s
+
+In addition, the "debug" prefix is treated specially for properties in Flutter.
+For the purposes of diagnostics, a property `foo` and a prefixed property
+`debugFoo` are treated as effectively describing the same property and it is
+sufficient to refer to one or the other.
+
+**BAD:**
+```
+class Absorber extends Widget {
+  bool get absorbing => _absorbing;
+  bool _absorbing;
+  bool get ignoringSemantics => _ignoringSemantics;
+  bool _ignoringSemantics;
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<bool>('absorbing', absorbing));
+    // Missing reference to ignoringSemantics
+  }
+}  
+```
+
+**GOOD:**
+```
+class Absorber extends Widget {
+  bool get absorbing => _absorbing;
+  bool _absorbing;
+  bool get ignoringSemantics => _ignoringSemantics;
+  bool _ignoringSemantics;
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<bool>('absorbing', absorbing));
+    properties.add(DiagnosticsProperty<bool>('ignoringSemantics', ignoringSemantics));
+  }
+}  
+```
+''';
+
+class DiagnosticsDescribeAllProperties extends LintRule
+    implements NodeLintRule {
+  DiagnosticsDescribeAllProperties()
+      : super(
+            name: 'diagnostic_describe_all_properties',
+            description: _desc,
+            details: _details,
+            maturity: Maturity.experimental,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = new _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+// todo (pq): for experiments and book-keeping; remove before landing
+int fileCount = 0;
+int debugPropertyCount = 0;
+int classesWithPropertiesButNoDebugFill = 0;
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  final InheritanceManager2 inheritanceManager;
+
+  _Visitor(this.rule, LinterContext context)
+      : inheritanceManager = new InheritanceManager2(context.typeSystem);
+
+  // todo (pq): for experiments and book-keeping; remove before landing
+  LineInfo lineInfo;
+
+  @override
+  visitCompilationUnit(CompilationUnit node) {
+    lineInfo = node.lineInfo;
+  }
+
+  static int noMethods = 0;
+  static int totalClasses = 0;
+
+  bool _isOverridingMember(Element member) {
+    if (member == null) {
+      return false;
+    }
+
+    ClassElement classElement =
+        member.getAncestor((element) => element is ClassElement);
+    if (classElement == null) {
+      return false;
+    }
+    Uri libraryUri = classElement.library.source.uri;
+    return inheritanceManager.getInherited(
+            classElement.type, new Name(libraryUri, member.name)) !=
```

Comment:
nit: trailing comma before the last `)` might also make this a bit more readable.

Reply:
```dart
    return inheritanceManager.getInherited(
            classElement.type,
            new Name(
              libraryUri,
              member.name,
            )) !=
        null;
```

😬

Reply:
or...

```dart
    return inheritanceManager.getInherited(
          classElement.type,
          new Name(libraryUri, member.name),
        ) !=
        null;
```

🤔